### PR TITLE
Prevent user changes to the voting config exclusions in atomic-register based coordination

### PR DESCRIPTION
**Pull request by fcofdez**

### CODE REVIEW

**Review by DaveCTurner — COMMENTED** [review]
This makes the voting config APIs into no-ops when using atomic-register-based masters, which is technically ok, but in fact we won't be hitting these APIs at all when using atomic-register-based masters. I'd rather we adjusted the test framework to more accurately reflect how we expect these things to be orchestrated, i.e. bypassing the voting-config-exclusions bits entirely.

**Comment by fcofdez** [comment]
I've adjusted the PR to disable the voting config exclusions through an overridable method in `ESIntegTestCase`. I couldn't find another way where we ensure that this flag is always false when we're running a cluster with atomic-register based coordination.

File: test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java [hunk 1]
```diff
@@ -2441,6 +2442,10 @@ protected static RestClient createRestClient(
      */
     protected void setupSuiteScopeCluster() throws Exception {}
 
+    protected boolean autoManageVotingExclusions() {
```
Comment (DaveCTurner) [inline test/framework/src/main/java/org/elasticsearch/test/ESIntegTestCase.java:2445]: Could we add a comment indicating that this is a workaround that we should remove once https://github.com/elastic/elasticsearch/issues/98055 is resolved?

File: server/src/main/java/org/elasticsearch/cluster/coordination/Reconfigurator.java [hunk 2]
```diff
@@ -148,6 +148,10 @@ public ClusterState maybeReconfigureAfterNewMasterIsElected(ClusterState cluster
         return clusterState;
     }
 
+    public void ensureVotingConfigCanBeModified() {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/Reconfigurator.java:151]: Could we add a comment indicating that this is a workaround that we should remove once https://github.com/elastic/elasticsearch/issues/98055 is resolved?

File: server/src/test/java/org/elasticsearch/action/admin/cluster/configuration/TransportAddVotingConfigExclusionsActionTests.java [hunk 3]
```diff
@@ -476,6 +481,28 @@ public void testTimesOut() throws InterruptedException {
...
+
+        reconfigurator.disableUserVotingConfigModifications();
+
+        transportService.sendRequest(
+            localNode,
+            AddVotingConfigExclusionsAction.NAME,
+            new AddVotingConfigExclusionsRequest(Strings.EMPTY_ARRAY, new String[] { "other1" }, TimeValue.timeValueMillis(100)),
+            expectError(e -> {
+                exceptionHolder.set(e);
+                countDownLatch.countDown();
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/action/admin/cluster/configuration/TransportAddVotingConfigExclusionsActionTests.java:496]: Matches the style of the other tests, but I opened https://github.com/elastic/elasticsearch/issues/98057 to remind us to improve this area

**Review by DaveCTurner — COMMENTED** [review]
Makes sense, I left a couple of suggestions for comments.

File: server/src/main/java/org/elasticsearch/cluster/coordination/stateless/SingleNodeReconfigurator.java [hunk 4]
```diff
@@ -49,4 +49,9 @@ public ClusterState maybeReconfigureAfterNewMasterIsElected(ClusterState cluster
             )
             .build();
     }
+
+    @Override
+    public void ensureVotingConfigCanBeModified() {
+        throw new IllegalStateException("Voting configuration cannot be modified using atomic-register based coordination");
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/coordination/stateless/SingleNodeReconfigurator.java:55]: Should we `assert false` here too? Calling these APIs is a bug I reckon.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**